src/main/java/com/android/volley/cronet/CronetHttpStack.java

```diff
@@ -0,0 +1,293 @@
+                builder.setHttpMethod("PUT");
+                addBodyIfExists(request.getBody(), builder);
+                break;
+            case Request.Method.HEAD:
+                builder.setHttpMethod("HEAD");
+                break;
+            case Request.Method.OPTIONS:
+                builder.setHttpMethod("OPTIONS");
+                break;
+            case Request.Method.TRACE:
+                builder.setHttpMethod("TRACE");
+                break;
+            case Request.Method.PATCH:
+                builder.setHttpMethod("PATCH");
+                addBodyIfExists(request.getBody(), builder);
+                break;
+            default:
+                throw new IllegalStateException("Unknown method type.");
+        }
+    }
+
+    /**
+     * Sets the request headers for the UrlRequest.
+     *
+     * @param request to get the headers from
+     * @param additionalHeaders for the UrlRequest
+     * @param builder that we are adding the request headers to
+     * @throws AuthFailureError is thrown if Request#getHeaders throws ones
+     */
+    private void setRequestHeaders(
+            Request<?> request, Map<String, String> additionalHeaders, UrlRequest.Builder builder)
+            throws AuthFailureError {
+        HashMap<String, String> map = new HashMap<>();
+        map.putAll(additionalHeaders);
+        // Request.getHeaders() takes precedence over the given additional (cache) headers).
+        map.putAll(request.getHeaders());
+        for (Map.Entry<String, String> header : map.entrySet()) {
+            builder.addHeader(header.getKey(), header.getValue());
+        }
+    }
+
+    /** Sets the UploadDataProvider of the UrlRequest.Builder */
+    private void addBodyIfExists(@Nullable byte[] body, UrlRequest.Builder builder) {
+        if (body != null) {
+            UploadDataProvider dataProvider = UploadDataProviders.create(body);
+            builder.setUploadDataProvider(dataProvider, mCallbackExecutor);
+        }
+    }
+
+    private int getContentLength(UrlResponseInfo urlResponseInfo) {
+        List<String> content = urlResponseInfo.getAllHeaders().get("Content-Length");
+        if (content == null) {
+            return 1024;
+        } else {
+            return Integer.parseInt(content.get(0));
+        }
+    }
+
+    /**
+     * Builder is used to build an instance of {@link CronetHttpStack} from values configured by the
+     * setters.
+     */
+    public static class Builder {
+        private static final int DEFAULT_POOL_SIZE = 4096;
+        private CronetEngine mCronetEngine;
+        private Executor mCallbackExecutor;
+        private ExecutorService mBlockingExecutor;
+        private ByteArrayPool mPool;
+        private UrlRewriter mUrlRewriter;
+
+        public Builder(
+                Context context, ExecutorService blockingExecutor, Executor callbackExecutor) {
+            mCronetEngine = new CronetEngine.Builder(context).build();
```

Comment:
My recommendation would be to defer this work to build() for any field not explicitly set. Otherwise, we're creating unnecessary CronetEngine, ByteArrayPool, and UrlRewriter objects that just get discarded.